src/core_plugins/choropleth/public/choropleth_controller.js

```diff
@@ -0,0 +1,156 @@
+
+      kibanaMap.useUiStateFromVisualization($scope.vis);
+      kibanaMap.resize();
+      $element.trigger('renderComplete');
+    });
+  });
+
+  $scope.$watch('vis.params', (visParams) => {
+    kibanaMapReady.then(() => {
+      if (!visParams.selectedJoinField) {
+        visParams.selectedJoinField = visParams.selectedLayer.fields[0];
+      }
+
+      updateChoroplethLayer(visParams.selectedLayer.url);
+      choroplethLayer.setJoinField(visParams.selectedJoinField.name);
+      choroplethLayer.setColorRamp(truncatedColorMaps[visParams.colorSchema]);
+      setTooltipFormatter();
+
+      kibanaMap.setShowTooltip(visParams.addTooltip);
+      kibanaMap.setLegendPosition(visParams.legendPosition);
+      kibanaMap.useUiStateFromVisualization($scope.vis);
+      kibanaMap.resize();
+      $element.trigger('renderComplete');
+    });
+  });
+
+  getAvailableVectorLayers()
+    .then(function (layersFromService) {
+      //todo: disable play-button until these settings are loaded!
+      $scope.vis.type.params.vectorLayers = $scope.vis.type.params.vectorLayers.concat(layersFromService);
+      $scope.$apply();
+    });
+
+  async function makeKibanaMap() {
+
+    const tmsSettings = await serviceSettings.getTMSService();
+    const minMaxZoom = tmsSettings.getMinMaxZoom(false);
+    kibanaMap = new KibanaMap(containerNode, minMaxZoom);
+    kibanaMap.setZoomLevel(2);
+    const url = tmsSettings.getUrl();
+    const options = tmsSettings.getTMSOptions();
+    kibanaMap.setBaseLayer({ baseLayerType: 'tms', options: { url, ...options } });
+    kibanaMap.addLegendControl();
+    kibanaMap.addFitControl();
+    kibanaMap.persistUiStateForVisualization($scope.vis);
+  }
+
+  function setTooltipFormatter() {
+    const metricsAgg = _.first($scope.vis.aggs.bySchemaName.metric);
+    if ($scope.vis.aggs.bySchemaName.segment && $scope.vis.aggs.bySchemaName.segment[0]) {
+      const fieldName = $scope.vis.aggs.bySchemaName.segment[0].makeLabel();
+      choroplethLayer.setTooltipFormatter(tooltipFormatter, metricsAgg, fieldName);
+    } else {
+      choroplethLayer.setTooltipFormatter(tooltipFormatter, metricsAgg, null);
+    }
+  }
+
+  function updateChoroplethLayer(url) {
+
+    if (choroplethLayer && choroplethLayer.equalsGeoJsonUrl(url)) {
+      return;
+    }
+    kibanaMap.removeLayer(choroplethLayer);
+
+    const previousMetrics = choroplethLayer ? choroplethLayer.getMetrics() : null;
+    const previousMetricsAgg = choroplethLayer ? choroplethLayer.getMetricsAgg() : null;
+    choroplethLayer = new ChoroplethLayer(url);
+    if (previousMetrics && previousMetricsAgg) {
+      choroplethLayer.setMetrics(previousMetrics, previousMetricsAgg);
+    }
+    choroplethLayer.on('select', function (event) {
+      const aggs = $scope.vis.aggs.getResponseAggs();
+      const aggConfigResult = new AggConfigResult(aggs[0], false, event, event);
+      $scope.vis.listeners.click({ point: { aggConfigResult: aggConfigResult } });
```

Comment:
not related to this PR .... but to the refactor ... do you want to write a "click" handler for adding filters to the filter bar which will not require creating AggConfigResults ?